fabric-transfer-api-v1/src/main/java/net/fabricmc/fabric/impl/transfer/item/ContainerComponentStorage.java

```diff
@@ -0,0 +1,184 @@
+		return parts.get(slot);
+	}
+
+	private class ContainerSlotWrapper implements SingleSlotStorage<ItemVariant> {
+		final int slot;
+
+		ContainerSlotWrapper(int slot) {
+			this.slot = slot;
+		}
+
+		private ItemStack getStack() {
+			List<ItemStack> stacks = ContainerComponentStorage.this.containerAccessor().fabric_getStacks();
+
+			if (stacks.size() <= slot) return ItemStack.EMPTY;
+
+			return stacks.get(slot);
+		}
+
+		protected boolean setStack(ItemStack stack, TransactionContext transaction) {
+			List<ItemStack> stacks = ContainerComponentStorage.this.container().stream().collect(Collectors.toList());
+
+			while (stacks.size() <= slot) stacks.add(ItemStack.EMPTY);
+
+			stacks.set(slot, stack);
+
+			ContainerItemContext ctx = ContainerComponentStorage.this.ctx;
+
+			ItemVariant newVariant = ctx.getItemVariant().withComponentChanges(ComponentChanges.builder()
+							.add(DataComponentTypes.CONTAINER, ContainerComponent.fromStacks(stacks))
+							.build());
+
+			return ctx.exchange(newVariant, 1, transaction) == 1;
+		}
+
+		@Override
+		public long insert(ItemVariant insertedVariant, long maxAmount, TransactionContext transaction) {
+			StoragePreconditions.notBlankNotNegative(insertedVariant, maxAmount);
+
+			ItemStack currentStack = getStack();
+
+			if ((insertedVariant.matches(currentStack) || currentStack.isEmpty()) && insertedVariant.getItem().canBeNested()) {
+				int insertedAmount = (int) Math.min(maxAmount, getCapacity() - currentStack.getCount());
+
+				if (insertedAmount > 0) {
+					currentStack = getStack().copy();
+
+					if (currentStack.isEmpty()) {
+						currentStack = insertedVariant.toStack(insertedAmount);
+					} else {
+						currentStack.increment(insertedAmount);
+					}
+
+					if (!setStack(currentStack, transaction)) return 0;
+
+					return insertedAmount;
+				}
+			}
+
+			return 0;
+		}
+
+		@Override
+		public long extract(ItemVariant variant, long maxAmount, TransactionContext transaction) {
+			StoragePreconditions.notBlankNotNegative(variant, maxAmount);
+
+			ItemStack currentStack = getStack();
+
+			if (variant.matches(currentStack)) {
+				int extracted = (int) Math.min(currentStack.getCount(), maxAmount);
+
+				if (extracted > 0) {
+					currentStack = getStack().copy();
+					currentStack.decrement(extracted);
+
+					if (!setStack(currentStack, transaction)) return 0;
+
+					return extracted;
+				}
+			}
+
+			return 0;
+		}
+
+		@Override
+		public boolean isResourceBlank() {
+			return getStack().isEmpty();
+		}
+
+		@Override
+		public ItemVariant getResource() {
+			return ItemVariant.of(getStack());
+		}
+
+		@Override
+		public long getAmount() {
+			return getStack().getCount();
+		}
+
+		@Override
+		public long getCapacity() {
+			return getStack().getItem().getMaxCount();
+		}
+
+		@Override
+		public String toString() {
+			return "ContainerSlotWrapper[%s#%d]".formatted(ContainerComponentStorage.this.ctx.getItemVariant().getRegistryEntry().getIdAsString(), slot);
```

Comment:
```suggestion
			return "ContainerSlotWrapper[%s#%d]".formatted(ContainerComponentStorage.this.ctx.getItemVariant(), slot);
```
`ItemVariant` already implements `toString()`.